usr/opt/OSC200/usr/share/man/man3/rand.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: rand.3,v $ $Revision: 3.2.5.3 $ (DEC) $Date: 1994/01/20 02:01:02 $
.\"
.\" (c) Copyright 1991, OPEN SOFTWARE FOUNDATION, INC.
.\" ALL RIGHTS RESERVED
.\" 
.EQ
delim $$
.EN
.TH rand 3
.SH NAME
.PP
\*Lrand\*O, \*Lrand_r\*O, \*Lsrand\*O \- Generates pseudo-random numbers
.SH LIBRARY
.PP
\*OStandard C Library (\*Llibc.a\*O),
.nL
Berkeley Compatibility Library (\*Llibbsd.a)\*O
.nL
Reentrant Library (\*Llibc_r.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <stdlib.h>
.sp .4v
\*Lint rand (void);
.sp .4v
\*Lint rand_r(
.nL
	unsigned int *\*Vseedptr\*L,
.nL
	int *\*Vrandval\*L );
.sp .4v
\*Lvoid srand (
.nL
	unsigned int\*O \*Vseed\*O);\*L
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vseed\*O"
\*OSpecifies an initial seed value.
.LI "\*Vseedptr\*O"
Points to a seed value, updated at each call.
.LI "\*Vrandval\*O"
Points to a place to store the random number.
.LE
.SH DESCRIPTION
.PP
.iX "rand function"
.iX "rand_r function"
.iX "srand function"
.iX "random numbers" "generating"
.iX "generating random numbers"
The
\*Lrand(\|)\*O
function
returns successive pseudo-random
numbers in the range from 0 (zero) to RAND_MAX.
The sequence of values returned depends on the seed value set with the
\*Lsrand(\|)\*O
function.
If
\*Lrand(\|)\*O
is called before any calls to
\*Lsrand(\|)\*O
have been made, the same sequence will be generated as when
\*Lsrand(\|)\*O
is first called with a seed value of 1.
.PP
The
\*Lrand_r(\|)\*O
function is the reentrant version of the
\*Lrand(\|)\*O
function, for use with multi-threaded applications.
The
\*Lrand_r(\|)\*O
function
places the seed value at the address pointed to by
\*Vseedptr\*O,
and places the random number at the address pointed to by
\*Vrandval\*O.
.PP
The \*Lsrand(\|)\*O function resets the random-number generator
to a random starting point. The generator is initially seeded with
a value of 1.
.PP
The
\*Lrand(\|)\*O function is a very simple
random-number generator. Its spectral properties, the mathematical
measurement of how random the number sequence is, are somewhat weak.
.PP
See the \*Ldrand48(\|)\*O and \*Lrandom(\|)\*O functions
for more elaborate random-number generators that have better spectral
properties.
.SH NOTES
.PP
The
\*Lrand(\|)\*O
function is not supported for multi-threaded applications.  Instead,
its reentrant equivalent
\*Lrand_r(\|)\*O
should be used with multiple threads.
.PP
The BSD version of the \*Lrand(\|)\*O function
.\" Begin Digital-specific information CORRECTION
.\" Changed the coding of the equation so the superscripts work per QAR 6711.
returns a number in the range 0 to (2^31)-1, rather than
0 to (2^15)-1, and can be used by compiling
.\" End Digital-specific information CORRECTION
the code with the Berkeley Compatibility Library (\*Llibbsd.a).\*O
.PP
There are better random number generators, as noted above; however,
the \*Lrand(\|)\*O and \*Lsrand(\|)\*O functions are the
interfaces defined for the ANSI C library.
.PP
The following functions define the semantics of the \*Lrand(\|)\*O
and \*Lsrand(\|)\*O functions, and are included here to
facilitate porting applications from different implementations:
.rS
\*Lstatic unsigned int next = 1;
.sp .4v
\*Lint rand(  )
.nL
\*L{
.sp .4v
	\*Lnext = next * 1103515245 + 12345;
.sp .4v
	\*Lreturn (  (next >>16) & RAND_MAX);
.nL
\*L}
.sp .4v
\*Lvoid srand (seed)
.nL
\*Lint seed;
.nL
\*L{
.sp .4v
\*Lnext = seed
.nL
\*L}
.rE
.VL 1.5i
.LI "\*LAES Support Level:\*O"
Full use (\*Lrand(\|)\*O, \*Lsrand(\|)\*O)
.LE
.SH RETURN VALUES
.PP
The
\*Lrand(\|)\*O
function returns the next pseudo-random number in the sequence.
.PP
Upon successful completion, the
\*Lrand_r(\|)\*O
function returns a value of 0 (zero).
Otherwise,
-1 is returned and
\*Lerrno\*O
is set to indicate the error.
.PP
The \*Lsrand(\|)\*O function returns no value.
.SH ERRORS
.PP
If the
\*Lrand_r(\|)\*O
function fails,
\*Lerrno\*O
may be set to the following value:
.VL 1i
.LI "[EINVAL]"
Either
\*Vseedptr\*O
or
\*Vrandval\*O
is a null pointer.
.EQ
delim off
.EN
.iX "rand function"
.iX "rand_r function"
.iX "srand function"
.LE
.SH RELATED INFORMATION
.PP
Functions:
\*Ldrand48\*O(3),
\*Lrandom\*O(3)
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/rand.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: rand.c,v $ $Revision: 4.2.5.2 $ (DEC) $Date: 1993/06/07 21:21:48 $";
#endif 
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
#if !defined(lint) && !defined(_NOIDENT)
static char rcsid[] = "@(#)$RCSfile: rand.c,v $ $Revision: 4.2.5.2 $ (OSF) $Date: 1993/06/07 21:21:48 $";
#endif
/*
 * FUNCTIONS: rand, srand
 *
 * (C) COPYRIGHT International Business Machines Corp. 1985, 1989 
 * All Rights Reserved
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 * Copyright (c) 1984 AT&T	
 * All Rights Reserved  
 *
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 *
 * rand.c	1.11  com/lib/c/gen,3.1,8943 9/9/89 09:27:25
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#if defined(_THREAD_SAFE)
#pragma weak rand_r = __rand_r
#endif
#endif
#include <stdlib.h>
#include <errno.h>

#include "ts_supp.h"

#ifdef _THREAD_SAFE
#include "rec_mutex.h"

extern struct rec_mutex _rand_rmutex;
#endif	/* _THREAD_SAFE */

#define	MAGIC(x)	(((x = x * 1103515245L + 12345)>>16) & RAND_MAX)

static long randx = 1;

/*
 * NAME:	srand
 *
 * FUNCTION:	The srand function uses the arguement as a seed for a new
 *		sequence of pseudo-random numbers to be returned by 
 *		subsequent calls to rand.
 *
 * RETURN VALUE DESCRIPTION:	
 *		returns no value
 *
 */
void	
srand(unsigned int seed)
{
	TS_LOCK(&_rand_rmutex);

	randx = seed;

	TS_UNLOCK(&_rand_rmutex);
}


/*
 * NAME:	rand
 *
 * FUNCTION:	The rand function computes a sequence of pseudo-random
 *		integers in the range 0 to RAND_MAX.
 *
 * RETURN VALUE DESCRIPTION:	
 *		returns a psuedo-random integer
 *
 */
int 
rand(void)
{
	int	retval;

	TS_LOCK(&_rand_rmutex);

	retval = MAGIC(randx);

	TS_UNLOCK(&_rand_rmutex);
	return (retval);
}


#ifdef _THREAD_SAFE
int
rand_r(unsigned int *seed, int *randval)
{
	if ((randval == NULL) || (seed == NULL)) {
		_Seterrno(EINVAL);
		return(-1);
	}
	*randval = MAGIC(*seed);

	return (0);
}
#endif	/* _THREAD_SAFE */
